File: sdk/typescript/src/signers.ts
/**
 * Cryptographic signers for GitClaw SDK.
 *
 * Supports Ed25519 and ECDSA P-256 signing algorithms.
 */

import * as ed from '@noble/ed25519';
import { p256 } from '@noble/curves/p256';
import { sha512 } from '@noble/hashes/sha512';
import { bytesToHex, hexToBytes } from '@noble/hashes/utils';
import * as fs from 'fs';

// Configure ed25519 to use sha512
ed.etc.sha512Sync = (...m) => sha512(ed.etc.concatBytes(...m));

/**
 * Abstract interface for cryptographic signers.
 */
export interface Signer {
  /**
   * Sign a message and return the signature bytes.
   */
  sign(message: Uint8Array): Uint8Array;

  /**
   * Return the public key in base64 format with type prefix.
   */
  publicKey(): string;

  /**
   * Verify a signature (for testing purposes).
   */
  verify(signature: Uint8Array, message: Uint8Array): boolean;
}

/**
 * Ed25519 digital signature implementation.
 */
export class Ed25519Signer implements Signer {
  private privateKey: Uint8Array;
  private pubKey: Uint8Array;

  /**
   * Initialize with an Ed25519 private key.
   *
   * @param privateKey - 32-byte Ed25519 private key seed
   */
  private constructor(privateKey: Uint8Array) {
    if (privateKey.length !== 32) {
      throw new Error(`Ed25519 private key must be 32 bytes, got ${privateKey.length}`);
    }
    this.privateKey = privateKey;
    this.pubKey = ed.getPublicKey(privateKey);
  }

  /**
   * Sign a message using Ed25519.
   *
   * @param message - The message bytes to sign
   * @returns 64-byte Ed25519 signature
   */
  sign(message: Uint8Array): Uint8Array {
    return ed.sign(message, this.privateKey);
  }

  /**
   * Return the public key as base64 with ed25519: prefix.
   *
   * @returns String in format "ed25519:<base64_public_key>"
   */
  publicKey(): string {
    return `ed25519:${Buffer.from(this.pubKey).toString('base64')}`;
  }

  /**
   * Return the raw 32-byte public key.
   */
  publicKeyBytes(): Uint8Array {
    return this.pubKey;
  }

  /**
   * Return the public key in PEM format.
   */
  publicKeyPem(): string {
    // Ed25519 public key in SubjectPublicKeyInfo format
    // OID for Ed25519: 1.3.101.112
    const oid = new Uint8Array([0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70]);
    const bitString = new Uint8Array([0x03, 0x21, 0x00, ...this.pubKey]);
    const spki = new Uint8Array([0x30, oid.length + bitString.length, ...oid, ...bitString]);

    const b64 = Buffer.from(spki).toString('base64');
    const lines = b64.match(/.{1,64}/g) || [];
    return `-----BEGIN PUBLIC KEY-----\n${lines.join('\n')}\n-----END PUBLIC KEY-----\n`;
  }

  /**
   * Return the private key in PEM format (for storage).
   */
  privateKeyPem(): string {
    // Ed25519 private key in PKCS#8 format
    // OID for Ed25519: 1.3.101.112
    const oid = new Uint8Array([0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70]);
    const privateKeyOctet = new Uint8Array([0x04, 0x20, ...this.privateKey]);
    const privateKeyInfo = new Uint8Array([0x04, privateKeyOctet.length, ...privateKeyOctet]);
    const version = new Uint8Array([0x02, 0x01, 0x00]);
    const pkcs8 = new Uint8Array([
      0x30,
      version.length + oid.length + privateKeyInfo.length,
      ...version,
      ...oid,
      ...privateKeyInfo,
    ]);

    const b64 = Buffer.from(pkcs8).toString('base64');
    const lines = b64.match(/.{1,64}/g) || [];
    return `-----BEGIN PRIVATE KEY-----\n${lines.join('\n')}\n-----END PRIVATE KEY-----\n`;
  }

  /**
   * Verify a signature (for testing purposes).
   *
   * @param signature - The signature to verify
   * @param message - The original message
   * @returns True if valid, false otherwise
   */
  verify(signature: Uint8Array, message: Uint8Array): boolean {
    try {
      return ed.verify(signature, message, this.pubKey);
    } catch {
      return false;
    }
  }

  /**
   * Load an Ed25519 signer from a PEM file.
   *
   * @param path - Path to PEM file containing Ed25519 private key
   * @returns Ed25519Signer instance
   */
  static fromPemFile(path: string): Ed25519Signer {
    const pemData = fs.readFileSync(path, 'utf-8');
    return Ed25519Signer.fromPem(pemData);
  }

  /**
   * Load an Ed25519 signer from a PEM string.
   *
   * @param pemString - PEM-encoded Ed25519 private key
   * @returns Ed25519Signer instance
   */
  static fromPem(pemString: string): Ed25519Signer {
    // Extract base64 content from PEM
    const lines = pemString.split('\n');
    const b64Lines = lines.filter(
      (line) => !line.startsWith('-----') && line.trim().length > 0
    );
    const b64 = b64Lines.join('');
    const der = Buffer.from(b64, 'base64');

    // Parse PKCS#8 structure to extract the 32-byte private key
    // PKCS#8 for Ed25519: SEQUENCE { version, algorithm, privateKey }
    // The privateKey is an OCTET STRING containing another OCTET STRING with the 32-byte key
    const privateKey = extractEd25519PrivateKey(der);
    return new Ed25519Signer(privateKey);
  }

  /**
   * Load an Ed25519 signer from raw 32-byte private key.
   *
   * @param keyBytes - 32-byte Ed25519 private key seed
   * @returns Ed25519Signer instance
   */
  static fromBytes(keyBytes: Uint8Array): Ed25519Signer {
    return new Ed25519Signer(keyBytes);
  }

  /**
   * Generate a new Ed25519 keypair.
   *
   * @returns Object with signer and publicKey string
   */
  static generate(): { signer: Ed25519Signer; publicKey: string } {
    const privateKey = ed.utils.randomPrivateKey();
    const signer = new Ed25519Signer(privateKey);
    return { signer, publicKey: signer.publicKey() };
  }
}

/**
 * ECDSA P-256 digital signature implementation.
 */
export class EcdsaSigner implements Signer {
  private privateKey: Uint8Array;
  private pubKey: Uint8Array;

  /**
   * Initialize with an ECDSA P-256 private key.
   *
   * @param privateKey - 32-byte ECDSA P-256 private key
   */
  private constructor(privateKey: Uint8Array) {
    if (privateKey.length !== 32) {
      throw new Error(`ECDSA P-256 private key must be 32 bytes, got ${privateKey.length}`);
    }
    this.privateKey = privateKey;
    // Get compressed public key (33 bytes)
    this.pubKey = p256.getPublicKey(privateKey, true);
  }

  /**
   * Sign a message using ECDSA P-256 with SHA-256.
   *
   * @param message - The message bytes to sign
   * @returns DER-encoded ECDSA signature
   */
  sign(message: Uint8Array): Uint8Array {
    const sig = p256.sign(message, this.privateKey);
    return sig.toDERRawBytes();
  }

  /**
   * Return the public key as base64 with ecdsa: prefix.
   *
   * @returns String in format "ecdsa:<base64_compressed_public_key>"
   */
  publicKey(): string {
    return `ecdsa:${Buffer.from(this.pubKey).toString('base64')}`;
  }

  /**
   * Return the compressed public key bytes.
   */
  publicKeyBytes(): Uint8Array {
    return this.pubKey;
  }

  /**
   * Return the public key in PEM format.
   */
  publicKeyPem(): string {
    // P-256 public key in SubjectPublicKeyInfo format
    // OID for P-256: 1.2.840.10045.3.1.7
    // OID for EC public key: 1.2.840.10045.2.1
    const uncompressedPubKey = p256.getPublicKey(this.privateKey, false);
    const algorithm = new Uint8Array([
      0x30,
      0x13,
      0x06,
      0x07,
      0x2a,
      0x86,
      0x48,
      0xce,
      0x3d,
      0x02,
      0x01, // ecPublicKey OID
      0x06,
      0x08,
      0x2a,
      0x86,
      0x48,
      0xce,
      0x3d,
      0x03,
      0x01,
      0x07, // P-256 OID
    ]);
    const bitString = new Uint8Array([0x03, uncompressedPubKey.length + 1, 0x00, ...uncompressedPubKey]);
    const spki = new Uint8Array([0x30, algorithm.length + bitString.length, ...algorithm, ...bitString]);

    const b64 = Buffer.from(spki).toString('base64');
    const lines = b64.match(/.{1,64}/g) || [];
    return `-----BEGIN PUBLIC KEY-----\n${lines.join('\n')}\n-----END PUBLIC KEY-----\n`;
  }

  /**
   * Return the private key in PEM format (for storage).
   */
  privateKeyPem(): string {
    // ECDSA private key in PKCS#8 format
    // Structure: SEQUENCE { version, algorithm, privateKey }
    // where privateKey is OCTET STRING containing EC private key

    // Build the inner EC private key (SEC1 format)
    const ecVersion = new Uint8Array([0x02, 0x01, 0x01]); // version 1
    const privateKeyOctet = new Uint8Array([0x04, 0x20, ...this.privateKey]);

    // EC private key without optional parameters (simpler format)
    const ecPrivateKeyLen = ecVersion.length + privateKeyOctet.length;
    const ecPrivateKey = new Uint8Array([0x30, ecPrivateKeyLen, ...ecVersion, ...privateKeyOctet]);

    // PKCS#8 wrapper
    const pkcs8Version = new Uint8Array([0x02, 0x01, 0x00]); // version 0
    const algorithm = new Uint8Array([
      0x30,
      0x13,
      0x06,
      0x07,
      0x2a,
      0x86,
      0x48,
      0xce,
      0x3d,
      0x02,
      0x01, // ecPublicKey OID
      0x06,
      0x08,
      0x2a,
      0x86,
      0x48,
      0xce,
      0x3d,
      0x03,
      0x01,
      0x07, // P-256 OID
    ]);
    const privateKeyInfo = new Uint8Array([0x04, ecPrivateKey.length, ...ecPrivateKey]);

    const totalLen = pkcs8Version.length + algorithm.length + privateKeyInfo.length;
    const pkcs8 = new Uint8Array([0x30, totalLen, ...pkcs8Version, ...algorithm, ...privateKeyInfo]);

    const b64 = Buffer.from(pkcs8).toString('base64');
    const lines = b64.match(/.{1,64}/g) || [];
    return `-----BEGIN PRIVATE KEY-----\n${lines.join('\n')}\n-----END PRIVATE KEY-----\n`;
  }

  /**
   * Verify a signature (for testing purposes).
   *
   * @param signature - The DER-encoded signature to verify
   * @param message - The original message
   * @returns True if valid, false otherwise
   */
  verify(signature: Uint8Array, message: Uint8Array): boolean {
    try {
      return p256.verify(signature, message, this.pubKey);
    } catch {
      return false;
    }
  }

  /**
   * Load an ECDSA signer from a PEM file.
   *
   * @param path - Path to PEM file containing ECDSA P-256 private key
   * @returns EcdsaSigner instance
   */
  static fromPemFile(path: string): EcdsaSigner {
    const pemData = fs.readFileSync(path, 'utf-8');
    return EcdsaSigner.fromPem(pemData);
  }

  /**
   * Load an ECDSA signer from a PEM string.
   *
   * @param pemString - PEM-encoded ECDSA P-256 private key
   * @returns EcdsaSigner instance
   */
  static fromPem(pemString: string): EcdsaSigner {
    // Extract base64 content from PEM
    const lines = pemString.split('\n');
    const b64Lines = lines.filter(
      (line) => !line.startsWith('-----') && line.trim().length > 0
    );
    const b64 = b64Lines.join('');
    const der = Buffer.from(b64, 'base64');

    // Parse PKCS#8 or SEC1 structure to extract the 32-byte private key
    const privateKey = extractEcdsaPrivateKey(der);
    return new EcdsaSigner(privateKey);
  }

  /**
   * Generate a new ECDSA P-256 keypair.
   *
   * @returns Object with signer and publicKey string
   */
  static generate(): { signer: EcdsaSigner; publicKey: string } {
    const privateKey = p256.utils.randomPrivateKey();
    const signer = new EcdsaSigner(privateKey);
    return { signer, publicKey: signer.publicKey() };
  }
}

/**
 * Extract Ed25519 private key from DER-encoded PKCS#8 structure.
 */
function extractEd25519PrivateKey(der: Uint8Array): Uint8Array {
  // PKCS#8 structure:
  // SEQUENCE {
  //   INTEGER (version = 0)
  //   SEQUENCE { OID (Ed25519) }
  //   OCTET STRING { OCTET STRING { 32-byte key } }
  // }

  let offset = 0;

  // Skip outer SEQUENCE tag and length
  if (der[offset] !== 0x30) throw new Error('Invalid PKCS#8: expected SEQUENCE');
  offset++;
  offset += readLength(der, offset).bytesRead;

  // Skip version INTEGER
  if (der[offset] !== 0x02) throw new Error('Invalid PKCS#8: expected INTEGER (version)');
  offset++;
  const versionLen = readLength(der, offset);
  offset += versionLen.bytesRead + versionLen.length;

  // Skip algorithm SEQUENCE
  if (der[offset] !== 0x30) throw new Error('Invalid PKCS#8: expected SEQUENCE (algorithm)');
  offset++;
  const algLen = readLength(der, offset);
  offset += algLen.bytesRead + algLen.length;

  // Read privateKey OCTET STRING
  if (der[offset] !== 0x04) throw new Error('Invalid PKCS#8: expected OCTET STRING');
  offset++;
  const outerOctetLen = readLength(der, offset);
  offset += outerOctetLen.bytesRead;

  // Read inner OCTET STRING containing the actual key
  if (der[offset] !== 0x04) throw new Error('Invalid PKCS#8: expected inner OCTET STRING');
  offset++;
  const innerOctetLen = readLength(der, offset);
  offset += innerOctetLen.bytesRead;

  if (innerOctetLen.length !== 32) {
    throw new Error(`Invalid Ed25519 key length: expected 32, got ${innerOctetLen.length}`);
  }

  return der.slice(offset, offset + 32);
}

/**
 * Extract ECDSA P-256 private key from DER-encoded PKCS#8 or SEC1 structure.
 */
function extractEcdsaPrivateKey(der: Uint8Array): Uint8Array {
  let offset = 0;

  // Check if this is PKCS#8 or SEC1 format
  if (der[offset] !== 0x30) throw new Error('Invalid key format: expected SEQUENCE');
  offset++;
  const outerLen = readLength(der, offset);
  offset += outerLen.bytesRead;

  // Check first element - if INTEGER, it's PKCS#8; if OCTET STRING, it's SEC1
  if (der[offset] === 0x02) {
    // PKCS#8 format
    // Skip version INTEGER
    offset++;
    const versionLen = readLength(der, offset);
    offset += versionLen.bytesRead + versionLen.length;

    // Skip algorithm SEQUENCE
    if (der[offset] !== 0x30) throw new Error('Invalid PKCS#8: expected SEQUENCE (algorithm)');
    offset++;
    const algLen = readLength(der, offset);
    offset += algLen.bytesRead + algLen.length;

    // Read privateKey OCTET STRING (contains SEC1 EC private key)
    if (der[offset] !== 0x04) throw new Error('Invalid PKCS#8: expected OCTET STRING');
    offset++;
    const octetLen = readLength(der, offset);
    offset += octetLen.bytesRead;

    // Now parse the SEC1 structure inside
    return extractSec1PrivateKey(der.slice(offset, offset + octetLen.length));
  } else {
    // SEC1 format (EC PRIVATE KEY)
    return extractSec1PrivateKey(der);
  }
}

/**
 * Extract private key from SEC1 EC private key structure.
 */
function extractSec1PrivateKey(der: Uint8Array): Uint8Array {
  let offset = 0;

  // SEQUENCE
  if (der[offset] !== 0x30) throw new Error('Invalid SEC1: expected SEQUENCE');
  offset++;
  readLength(der, offset);
  offset += readLength(der, offset).bytesRead;

  // version INTEGER (should be 1)
  if (der[offset] !== 0x02) throw new Error('Invalid SEC1: expected INTEGER (version)');
  offset++;
  const versionLen = readLength(der, offset);
  offset += versionLen.bytesRead + versionLen.length;

  // privateKey OCTET STRING
  if (der[offset] !== 0x04) throw new Error('Invalid SEC1: expected OCTET STRING');
  offset++;
  const keyLen = readLength(der, offset);
  offset += keyLen.bytesRead;

  if (keyLen.length !== 32) {
    throw new Error(`Invalid ECDSA P-256 key length: expected 32, got ${keyLen.length}`);
  }

  return der.slice(offset, offset + 32);
}

/**
 * Read ASN.1 length field.
 */
function readLength(der: Uint8Array, offset: number): { length: number; bytesRead: number } {
  const firstByte = der[offset];

  if (firstByte < 0x80) {
    // Short form
    return { length: firstByte, bytesRead: 1 };
  }

  // Long form
  const numBytes = firstByte & 0x7f;
  let length = 0;
  for (let i = 0; i < numBytes; i++) {
    length = (length << 8) | der[offset + 1 + i];
  }
  return { length, bytesRead: 1 + numBytes };
}

// Re-export utility functions for testing
export { bytesToHex, hexToBytes };

File: sdk/rust/src/signers.rs
//! Cryptographic signers for `GitClaw` SDK.
//!
//! Supports Ed25519 and ECDSA P-256 signing algorithms.
//!
//! # Example
//!
//! ```rust
//! use gitclaw::{Ed25519Signer, Signer};
//!
//! // Generate a new keypair
//! let (signer, public_key) = Ed25519Signer::generate();
//! println!("Public key: {}", public_key);
//!
//! // Sign a message
//! let message = b"Hello, GitClaw!";
//! let signature = signer.sign(message).unwrap();
//! ```

use std::path::Path;

use base64::{engine::general_purpose::STANDARD as BASE64, Engine};
use ed25519_dalek::{
    Signature as Ed25519Signature, Signer as DalekSigner, SigningKey, Verifier, VerifyingKey,
};
use p256::ecdsa::{
    signature::Signer as EcdsaSignerTrait, signature::Verifier as EcdsaVerifierTrait,
    Signature as P256Signature, SigningKey as P256SigningKey, VerifyingKey as P256VerifyingKey,
};
use p256::pkcs8::{DecodePrivateKey, EncodePrivateKey, EncodePublicKey, LineEnding};
use rand::rngs::OsRng;

use crate::error::Error;

/// Trait for cryptographic signers.
pub trait Signer: Send + Sync {
    /// Sign a message and return the signature bytes.
    ///
    /// # Errors
    ///
    /// Returns an error if signing fails.
    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, Error>;

    /// Return the public key in base64 format with type prefix.
    fn public_key(&self) -> String;

    /// Verify a signature (for testing purposes).
    fn verify(&self, signature: &[u8], message: &[u8]) -> bool;
}

/// Ed25519 digital signature implementation.
pub struct Ed25519Signer {
    signing_key: SigningKey,
    verifying_key: VerifyingKey,
}

impl Ed25519Signer {
    /// Create a new Ed25519 signer from a signing key.
    fn new(signing_key: SigningKey) -> Self {
        let verifying_key = signing_key.verifying_key();
        Self {
            signing_key,
            verifying_key,
        }
    }

    /// Generate a new Ed25519 keypair.
    ///
    /// Returns a tuple of (signer, `public_key_string`).
    #[must_use]
    pub fn generate() -> (Self, String) {
        let signing_key = SigningKey::generate(&mut OsRng);
        let signer = Self::new(signing_key);
        let public_key = signer.public_key();
        (signer, public_key)
    }

    /// Load an Ed25519 signer from a PEM file.
    ///
    /// # Errors
    ///
    /// Returns an error if the file cannot be read or the key is invalid.
    pub fn from_pem_file<P: AsRef<Path>>(path: P) -> Result<Self, Error> {
        let pem_data = std::fs::read_to_string(path)?;
        Self::from_pem(&pem_data)
    }

    /// Load an Ed25519 signer from a PEM string.
    ///
    /// # Errors
    ///
    /// Returns an error if the PEM is invalid or not an Ed25519 key.
    pub fn from_pem(pem_string: &str) -> Result<Self, Error> {
        // Parse PEM
        let pem = pem::parse(pem_string)?;

        // Ed25519 private keys in PKCS#8 format
        if pem.tag() != "PRIVATE KEY" {
            return Err(Error::Key(format!(
                "Expected PRIVATE KEY, got {}",
                pem.tag()
            )));
        }

        // Parse PKCS#8 DER
        let signing_key = SigningKey::from_pkcs8_der(pem.contents())
            .map_err(|e| Error::Key(format!("Failed to parse Ed25519 key: {e}")))?;

        Ok(Self::new(signing_key))
    }

    /// Load an Ed25519 signer from raw 32-byte private key seed.
    ///
    /// # Errors
    ///
    /// This function is infallible for valid 32-byte arrays.
    pub fn from_bytes(key_bytes: &[u8; 32]) -> Result<Self, Error> {
        let signing_key = SigningKey::from_bytes(key_bytes);
        Ok(Self::new(signing_key))
    }

    /// Return the raw 32-byte public key.
    #[must_use]
    pub fn public_key_bytes(&self) -> [u8; 32] {
        self.verifying_key.to_bytes()
    }

    /// Return the public key in PEM format.
    #[must_use]
    pub fn public_key_pem(&self) -> String {
        let public_bytes = self.verifying_key.to_bytes();
        // Ed25519 public key in SubjectPublicKeyInfo format
        // OID: 1.3.101.112
        let mut der = vec![
            0x30, 0x2a, // SEQUENCE, 42 bytes
            0x30, 0x05, // SEQUENCE, 5 bytes (algorithm identifier)
            0x06, 0x03, 0x2b, 0x65, 0x70, // OID 1.3.101.112 (Ed25519)
            0x03, 0x21, 0x00, // BIT STRING, 33 bytes (1 unused bit byte + 32 key bytes)
        ];
        der.extend_from_slice(&public_bytes);

        let pem = pem::Pem::new("PUBLIC KEY", der);
        pem::encode(&pem)
    }

    /// Return the private key in PEM format (for storage).
    ///
    /// # Panics
    ///
    /// Panics if the key cannot be encoded (should never happen for valid keys).
    #[must_use]
    #[allow(clippy::expect_used)]
    pub fn private_key_pem(&self) -> String {
        // Use PKCS#8 format
        let pkcs8_der = self
            .signing_key
            .to_pkcs8_der()
            .expect("Failed to encode private key");
        let pem = pem::Pem::new("PRIVATE KEY", pkcs8_der.as_bytes().to_vec());
        pem::encode(&pem)
    }
}

impl Signer for Ed25519Signer {
    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, Error> {
        let signature: Ed25519Signature = self.signing_key.sign(message);
        Ok(signature.to_bytes().to_vec())
    }

    fn public_key(&self) -> String {
        let public_bytes = self.verifying_key.to_bytes();
        format!("ed25519:{}", BASE64.encode(public_bytes))
    }

    fn verify(&self, signature: &[u8], message: &[u8]) -> bool {
        if signature.len() != 64 {
            return false;
        }
        let Ok(sig_bytes): Result<[u8; 64], _> = signature.try_into() else {
            return false;
        };
        let sig = Ed25519Signature::from_bytes(&sig_bytes);
        self.verifying_key.verify(message, &sig).is_ok()
    }
}

/// ECDSA P-256 digital signature implementation.
pub struct EcdsaSigner {
    signing_key: P256SigningKey,
    verifying_key: P256VerifyingKey,
}

impl EcdsaSigner {
    /// Create a new ECDSA signer from a signing key.
    fn new(signing_key: P256SigningKey) -> Self {
        let verifying_key = *signing_key.verifying_key();
        Self {
            signing_key,
            verifying_key,
        }
    }

    /// Generate a new ECDSA P-256 keypair.
    ///
    /// Returns a tuple of (signer, `public_key_string`).
    #[must_use]
    pub fn generate() -> (Self, String) {
        let signing_key = P256SigningKey::random(&mut OsRng);
        let signer = Self::new(signing_key);
        let public_key = signer.public_key();
        (signer, public_key)
    }

    /// Load an ECDSA signer from a PEM file.
    ///
    /// # Errors
    ///
    /// Returns an error if the file cannot be read or the key is invalid.
    pub fn from_pem_file<P: AsRef<Path>>(path: P) -> Result<Self, Error> {
        let pem_data = std::fs::read_to_string(path)?;
        Self::from_pem(&pem_data)
    }

    /// Load an ECDSA signer from a PEM string.
    ///
    /// # Errors
    ///
    /// Returns an error if the PEM is invalid or not an ECDSA P-256 key.
    pub fn from_pem(pem_string: &str) -> Result<Self, Error> {
        let signing_key = P256SigningKey::from_pkcs8_pem(pem_string)
            .map_err(|e| Error::Key(format!("Failed to parse ECDSA key: {e}")))?;

        Ok(Self::new(signing_key))
    }

    /// Return the compressed public key bytes (33 bytes).
    #[must_use]
    pub fn public_key_bytes(&self) -> Vec<u8> {
        #[allow(unused_imports)]
        use p256::elliptic_curve::sec1::ToEncodedPoint as _;
        // Use SEC1 compressed point encoding (33 bytes for P-256)
        self.verifying_key
            .to_encoded_point(true)
            .as_bytes()
            .to_vec()
    }

    /// Return the public key in PEM format.
    ///
    /// # Panics
    ///
    /// Panics if the key cannot be encoded (should never happen for valid keys).
    #[must_use]
    #[allow(clippy::expect_used)]
    pub fn public_key_pem(&self) -> String {
        self.verifying_key
            .to_public_key_pem(LineEnding::default())
            .expect("Failed to encode public key")
    }

    /// Return the private key in PEM format (for storage).
    ///
    /// # Panics
    ///
    /// Panics if the key cannot be encoded (should never happen for valid keys).
    #[must_use]
    #[allow(clippy::expect_used)]
    pub fn private_key_pem(&self) -> String {
        self.signing_key
            .to_pkcs8_pem(LineEnding::default())
            .expect("Failed to encode private key")
            .to_string()
    }
}

impl Signer for EcdsaSigner {
    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, Error> {
        let signature: P256Signature = EcdsaSignerTrait::sign(&self.signing_key, message);
        Ok(signature.to_der().as_bytes().to_vec())
    }

    fn public_key(&self) -> String {
        let public_bytes = self.public_key_bytes();
        format!("ecdsa:{}", BASE64.encode(public_bytes))
    }

    fn verify(&self, signature: &[u8], message: &[u8]) -> bool {
        let Ok(sig) = P256Signature::from_der(signature) else {
            return false;
        };
        EcdsaVerifierTrait::verify(&self.verifying_key, message, &sig).is_ok()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ed25519_generate_and_sign() {
        let (signer, public_key) = Ed25519Signer::generate();
        assert!(public_key.starts_with("ed25519:"));

        let message = b"test message";
        let signature = signer.sign(message).unwrap();
        assert_eq!(signature.len(), 64);
        assert!(signer.verify(&signature, message));
    }

    #[test]
    fn test_ed25519_signature_is_deterministic() {
        let (signer, _) = Ed25519Signer::generate();
        let message = b"test message";

        let sig1 = signer.sign(message).unwrap();
        let sig2 = signer.sign(message).unwrap();

        assert_eq!(sig1, sig2, "Ed25519 signatures should be deterministic");
    }

    #[test]
    fn test_ed25519_pem_round_trip() {
        let (signer, public_key) = Ed25519Signer::generate();
        let pem = signer.private_key_pem();

        let loaded = Ed25519Signer::from_pem(&pem).unwrap();
        assert_eq!(loaded.public_key(), public_key);

        // Sign with both and verify
        let message = b"test message";
        let sig1 = signer.sign(message).unwrap();
        let sig2 = loaded.sign(message).unwrap();
        assert_eq!(sig1, sig2);
    }

    #[test]
    fn test_ed25519_from_bytes() {
        let seed: [u8; 32] = [42; 32];
        let signer1 = Ed25519Signer::from_bytes(&seed).unwrap();
        let signer2 = Ed25519Signer::from_bytes(&seed).unwrap();

        assert_eq!(
            signer1.public_key(),
            signer2.public_key(),
            "Same seed should produce same key"
        );
    }

    #[test]
    fn test_ecdsa_generate_and_sign() {
        let (signer, public_key) = EcdsaSigner::generate();
        assert!(public_key.starts_with("ecdsa:"));

        let message = b"test message";
        let signature = signer.sign(message).unwrap();
        // DER-encoded P-256 signatures are typically 70-72 bytes
        assert!((68..=72).contains(&signature.len()));
        assert!(signer.verify(&signature, message));
    }

    #[test]
    fn test_ecdsa_pem_round_trip() {
        let (signer, public_key) = EcdsaSigner::generate();
        let pem = signer.private_key_pem();

        let loaded = EcdsaSigner::from_pem(&pem).unwrap();
        assert_eq!(loaded.public_key(), public_key);

        // Sign with loaded and verify with original
        let message = b"test message";
        let sig = loaded.sign(message).unwrap();
        assert!(signer.verify(&sig, message));
    }

    #[test]
    fn test_ecdsa_public_key_is_compressed() {
        let (_signer, public_key) = EcdsaSigner::generate();

        // Extract base64 part
        let b64_part = &public_key[6..]; // Remove "ecdsa:" prefix
        let decoded = BASE64.decode(b64_part).unwrap();

        // Compressed P-256 public key is 33 bytes
        assert_eq!(decoded.len(), 33);
    }
}
